src/components/stateful-enroll/data/hooks/useStatefulEnroll.ts

```diff
@@ -0,0 +1,212 @@
+import {
+  useCallback, useContext, useEffect, useState,
+} from 'react';
+import { queryOptions, useMutation, useQuery } from '@tanstack/react-query';
+import { AppContext } from '@edx/frontend-platform/react';
+import { logError, logInfo } from '@edx/frontend-platform/logging';
+import { useOptimizelyEnrollmentClickHandler, useTrackSearchConversionClickHandler } from '../../../course/data/hooks';
+import { EVENT_NAMES } from '../../../course/data/constants';
+import { queryPolicyTransaction, useEnterpriseCustomer } from '../../../app/data';
+
+import { submitRedemptionRequest } from '../service';
+
+interface BaseArgs {
+  contentKey: string;
+  subsidyAccessPolicy?: SubsidyAccessPolicy;
+  onBeginRedeem?: () => void;
+  onSuccess: (transaction: SubsidyTransaction) => void;
+  onError: (error: Error) => void;
+}
+
+interface UseStatefulEnrollArgs extends BaseArgs {
+  userEnrollments: EnterpriseCourseEnrollment[];
+}
+
+interface UseRedemptionArgs extends BaseArgs {
+  resetTransaction: () => void;
+}
+
+interface UseTransactionStatusArgs {
+  transaction?: SubsidyTransaction;
+  onSuccess: (transaction: SubsidyTransaction) => void;
+  onError: (error: Error) => void;
+}
+
+interface RedeemFnArgs {
+  metadata?: Record<string, unknown>;
+}
+
+/**
+ * Returns whether the transaction state should be polled (i.e., the transaction
+ * is pending).
+ */
+function shouldPollTransactionState(transaction?: SubsidyTransaction) {
```

Comment:
When a user selects enroll, do we expect the UI of the button to update to "Enrolling..."?
Based on the code it seems like its supposed to. ([src](https://github.com/openedx/frontend-app-learner-portal-enterprise/blob/ab301319d772f54623c3ad453f15e6fcae8f2629/src/components/stateful-enroll/StatefulEnroll.jsx#L50-L57))

It looks like it does not do that currently and is currently the state on stage at the moment.

From local testing of PR:
https://github.com/user-attachments/assets/1f7e0c90-13ef-4748-ade5-95c5c738b5c0

Reply:
IIRC, we discussed this synchronously. The screen recording shows the applicable subsidy is a subscription, so the "Enroll" CTA is a standard hyperlink to the DSC flow vs. an asynchronous `StatefulButton` for learner credit redemptions.